inst/stan/functions/convolve.stan

```diff
@@ -1,48 +1,88 @@
+vector convolve_dot_product(vector x, vector y, int len) {
+    int ylen = num_elements(y);
+    vector[len] z;
+    for (s in 1:len) {
+        z[s] = 1e-5 + // add a small value to ensure numerical stability
+          dot_product(
+            x[max(1, (s - ylen + 1)):s], tail(y, min(ylen, s))
+          );
+    }
+   return(z);
   }
 
 
 // convolve latent infections to reported (but still unobserved) cases
 vector convolve_to_report(vector infections,
-                          real[] delay_mean,
-                          real[] delay_sd,
-                          int[] max_delay,
+                          vector delay_pmf,
                           int seeding_time) {
   int t = num_elements(infections);
   vector[t - seeding_time] reports;
   vector[t] unobs_reports = infections;
-  int delays = num_elements(delay_mean);
+  int delays = num_elements(delay_pmf);
   if (delays) {
-    for (s in 1:delays) {
-      vector[max_delay[s]] pmf;
-      pmf = discretised_lognormal_pmf(
-        delay_mean[s], delay_sd[s], max_delay[s], 1
-      );
-      unobs_reports = convolve(unobs_reports, pmf);
-    }
+    unobs_reports = convolve_dot_product(unobs_reports, delay_pmf, t);
     reports = unobs_reports[(seeding_time + 1):t];
-  }else{
+  } else {
     reports = infections[(seeding_time + 1):t];
   }
   return(reports);
 }
 
 void delays_lp(real[] delay_mean, real[] delay_mean_mean, real[] delay_mean_sd,
                real[] delay_sd, real[] delay_sd_mean, real[] delay_sd_sd, int weight){
-    int delays = num_elements(delay_mean);
-    if (delays) {
-      for (s in 1:delays) {
-       target += normal_lpdf(delay_mean[s] | delay_mean_mean[s], delay_mean_sd[s]) * weight;
-       target += normal_lpdf(delay_sd[s] | delay_sd_mean[s], delay_sd_sd[s]) * weight;
+    int mean_delays = num_elements(delay_mean);
+    int sd_delays = num_elements(delay_sd);
+    if (mean_delays) {
+      for (s in 1:mean_delays) {
+        if (delay_mean_sd[s] > 0) {
+          // uncertain mean
+          target += normal_lpdf(delay_mean[s] | delay_mean_mean[s], delay_mean_sd[s]) * weight;
+        }
+      }
+    }
+    if (sd_delays) {
+      for (s in 1:sd_delays) {
+        if (delay_sd_sd[s] > 0) {
+          // uncertain sd
+          target += normal_lpdf(delay_sd[s] | delay_sd_mean[s], delay_sd_sd[s]) * weight;
+        }
      }
   }
 }
+
+// combined fixed/variable pmfs
+vector combine_pmfs(vector fixed_pmf, real[] pmf_mu, real[] pmf_sigma, int[] pmf_n, int[] dist, int len, int truncate) {
+  int n_fixed = num_elements(fixed_pmf);
+  int n_variable = num_elements(pmf_mu);
+  vector[len] pmf = rep_vector(0, len);
+  if (n_fixed > 0) {
+    pmf[1:n_fixed] = fixed_pmf;
+  } else if (n_variable > 0) {
+    pmf[1] = 1;
+  }
+  for (s in 1:n_variable) {
+    vector[pmf_n[s]] variable_pmf;
+    variable_pmf = discretised_pmf(pmf_mu[s], pmf_sigma[s], pmf_n[s], dist[s], truncate);
+    pmf = convolve(pmf, variable_pmf, len);
```

Comment:
For delay distributions we don't appear to be reversing the pmf anywhere. I'd suggest we add to here vs having in each function.

Reply:
I've surfaced these all into `combine_pmfs` which comes with a slight reduction in overhears but perhaps slightly harder to read code.

Reply:
Perhaps we could call `_pmf` variables `_rev_pmf` when they're reversed? I found it quite hard to keep track of what was forward and what backward when reading the code in the past.

Reply:
> I'd suggest we add to here vs having in each function.

Is this for readability or efficiency? Personally I found it easier to follow the model when `pmf` definitions were forward but that might of course just me me.

Reply:
its more efficient but also because it centralises where these transformations happen. Unless I missed something the changes in this branch led to delay distributions not being reversed and I think that was because there wasn't water tight logic on where this transformation should be happening. 

Yes can update the naming scheme.